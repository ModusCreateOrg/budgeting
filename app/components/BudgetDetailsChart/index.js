// @flow
import * as React from 'react';
import type { Transaction } from 'modules/transactions';
import { scaleQuantile } from 'd3';
import DonutChart from 'components/DonutChart';
import styles from './style.scss';

type BudgetDetailsChartProps = {
  transaction: Transaction,
  flow: Object,
};

const BudgetDetailsChart = ({ transaction, flow }: BudgetDetailsChartProps) => {
  const { id, value, description } = transaction;
  const isNegative = value < 0;
  const typeOfFlow = flow[isNegative ? 'outflow' : 'inflow'];
  const percent = (value / typeOfFlow * 100).toFixed(2);
  const amountCls = isNegative ? styles.neg : styles.pos;
  const colorRange = isNegative ? ['crimson', 'orange'] : ['lightgreen', 'steelblue'];
  const color = scaleQuantile().range(colorRange);
  const chartData = [
    { id, value: Math.abs(value), description },
    { id: id * -1, value: Math.abs(typeOfFlow - value), description: `Remaining ${isNegative ? 'Outflow' : 'Inflow'}` },
  ];

  return (
    <section className={styles.chartContainer}>
      <h1 className={styles.title}>{description}</h1>
      <h2 className={amountCls}>{`${isNegative ? '-' : '+'}${percent}%`}</h2>
      <DonutChart
        dataKey="id"
        data={chartData}
        dataLabel="description"
        dataValue="value"
        innerRatio={Math.abs(typeOfFlow)}
        color={color}
      />
    </section>
  );
};

export default BudgetDetailsChart;
