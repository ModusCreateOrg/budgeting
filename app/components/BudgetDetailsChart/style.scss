@import 'theme/variables';

.chartContainer {
    display: flex;
    align-items: center;
    flex-direction: column;

    h1, h2 {
        margin: 0;
    }
    > div > svg {
        @media only screen and (max-width: $screen-small) {
            flex: none;
            min-height: 200px;
            height: calc(100% - 100px);
            width: 100%;
        }
    }
}

.title {
    text-transform: capitalize;
    text-align: center;
}

.neg {
    color: $red;
}
.pos {
    color: $green;
}