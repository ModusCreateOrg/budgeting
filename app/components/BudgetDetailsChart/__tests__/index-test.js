import React from 'react';
import renderer from 'react-test-renderer';
import BudgetDetailsChart from 'components/BudgetDetailsChart';

it('renders correctly', () => {
  const mockTransaction = {
    id: 1,
    description: "Trader Joe's food",
    value: -423.34,
    categoryId: 1,
  };

  const mockFlow = {
    inflow: 14000,
    outflow: -2000,
  };

  const tree = renderer.create(<BudgetDetailsChart transaction={mockTransaction} flow={mockFlow} />).toJSON();
  expect(tree).toMatchSnapshot();
});
