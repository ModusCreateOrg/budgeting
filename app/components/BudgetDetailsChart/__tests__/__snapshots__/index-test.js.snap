// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  className="chartContainer"
>
  <h1
    className="title"
  >
    Trader Joe's food
  </h1>
  <h2
    className="neg"
  >
    -21.17%
  </h2>
  <div
    className="donutChart"
  >
    <svg
      className="mainSvg"
      height={316}
      viewBox="-8 -8 316 316"
      width={316}
    >
      <g
        transform="translate(150,150)"
      >
        <path
          d="M9.184850993605149e-15,-150A150,150,0,0,1,145.6708913294543,-35.77696772062049L0.1456708913294543,-0.035776967720620494A0.15,0.15,0,0,0,9.184850993605149e-18,-0.15Z"
          fill="crimson"
        />
        <path
          d="M145.6708913294543,-35.77696772062049A150,150,0,1,1,1.0567220997420333e-13,-150L1.0567220997420333e-16,-0.15A0.15,0.15,0,1,0,0.1456708913294543,-0.035776967720620494Z"
          fill="orange"
        />
      </g>
    </svg>
    <ul
      className="legend"
    >
      <li
        style={
          Object {
            "color": "crimson",
          }
        }
      >
        <span>
          Trader Joe's food
        </span>
        <span
          className="value"
        >
           
          $423.34
           
        </span>
      </li>
      <li
        style={
          Object {
            "color": "orange",
          }
        }
      >
        <span>
          Remaining Outflow
        </span>
        <span
          className="value"
        >
           
          $1,576.66
           
        </span>
      </li>
    </ul>
  </div>
</section>
`;
